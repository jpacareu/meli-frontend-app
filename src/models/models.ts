import { History } from 'history';

export interface IState {
	items: IItem[];
	searchBox: string;
	categories: string[];
	loadingSearch: boolean;
	detailProduct: {
		item: IDetailItem;
	} | null;
}

export interface IActions {
	search: (...rest: any) => {};
	getProductDetail: (...rest: any) => {};
	loading: (...rest: any) => {};
	received: (...rest: any) => {};
	setState: (...rest: any) => {};
	handleError: (...rest: any) => {};
}

export interface IStateHistory extends IState {
	history: History;
}

export interface IDetailItem {
	title: string;
	price: IPrice;
	picture: string;
	condition: ICondition;
	free_shipping: boolean;
	sold_quantity: number;
	description: string;
}

export interface IItem {
	id: string;
	title: string;
	price: IPrice;
	picture: string;
	condition: ICondition;
	free_shipping: boolean;
}

export type ICondition = 'new' | 'used';

export interface IPrice {
	currency: string;
	amount: number;
	decimals: number;
}

export interface IStore {
	Provider: any;
	connect: () => any;
	actions: IActions;
}
