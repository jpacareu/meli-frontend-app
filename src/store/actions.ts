import axios from 'axios';
import { IState, IActions } from '../models/models';

const baseURL =
	process.env.NODE_ENV === 'production'
		? 'https://meli-backend-app.herokuapp.com'
		: 'http://localhost:4000';

const axiosInstance = axios.create({
	baseURL: baseURL + '/api/'
});

export const actionsCreators: IActions = {
	search: async (state: IState, actions) => {
		let resultState: any = {};

		if (state.searchBox === '') return resultState;

		await actions.loading();

		try {
			const result = await axiosInstance.get('/items', {
				params: {
					q: state.searchBox
				}
			});

			const { items, categories } = result.data;
			resultState = { items: items || [], categories: categories || [] };
		} catch (error) {
			await actions.handleError(error);
		} finally {
			await actions.received();
		}

		return resultState;
	},
	getProductDetail: async (state: IState, actions, id: string = '') => {
		let resultState: any = {};
		await actions.loading();
		try {
			const result = await axiosInstance.get(`/items/${id}`);
			const { categories, ...rest } = result.data;
			resultState = {
				detailProduct: rest,
				categories
			};
		} catch (error) {
			await actions.handleError(error);
		} finally {
			await actions.received();
			return resultState;
		}
	},
	loading: () => {
		return { loadingSearch: true };
	},
	received: () => {
		return { loadingSearch: false };
	},
	handleError: (state, actions, error) => {
		// Handle error mock
		return {};
	},
	setState: (state, actions, value = {}) => {
		return { ...value };
	}
};
