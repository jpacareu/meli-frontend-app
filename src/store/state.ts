import { IState } from '../models/models';

export const initialState: IState = {
	items: [],
	searchBox: '',
	categories: [],
	loadingSearch: false,
	detailProduct: null
};
