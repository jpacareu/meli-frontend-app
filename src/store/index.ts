import createStore from 'react-waterfall';
import { actionsCreators } from './actions';
import { initialState } from './state';

const config = {
	initialState: initialState,
	actionsCreators: actionsCreators
};
const store = createStore(config);

export const { Provider, connect, actions } = store;
