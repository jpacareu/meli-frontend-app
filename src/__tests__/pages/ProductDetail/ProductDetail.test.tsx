import React from 'react';
import renderer, { ReactTestRenderer } from 'react-test-renderer';
import ProductDetailDefault, {
	ProductDetail
} from '../../../pages/ProductDetail/ProductDetail';
import Spinner from '../../../components/Spinner';
import { IStateHistory } from '../../../models/models';
import { Provider } from '../../../store';
import { match } from 'react-router';
import { History } from 'history';

type IMatch = match<{ id: string }>;

interface ITestProps extends IStateHistory {
	match: IMatch;
}

const props: ITestProps = {
	match: {
		params: {
			id: '123'
		}
	} as IMatch,
	loadingSearch: false,
	detailProduct: {
		item: {
			condition: 'new',
			description: 'This is the description',
			free_shipping: true,
			picture: 'http://mywebsite.com/image.jpg',
			sold_quantity: 1000,
			title: 'Title 1',
			price: {
				amount: 1000,
				currency: 'ARS',
				decimals: 0
			}
		}
	},
	categories: ['Category 1', 'Category 2', 'Category 3'],
	items: [],
	searchBox: '',
	history: {} as History
};

let component: JSX.Element | null = null;
let tree: ReactTestRenderer;

describe('<ProductDetailDefault />', () => {
	beforeEach(() => {
		component = (
			<Provider>
				<ProductDetailDefault {...props} />
			</Provider>
		);
		tree = renderer.create(component);
	});

	it('should render without crashing', () => {
		expect(tree.toJSON()).toMatchSnapshot();
	});
});

describe('<ProductDetail />', () => {
	beforeEach(() => {
		component = (
			<Provider>
				<ProductDetail {...props} />
			</Provider>
		);
		tree = renderer.create(component);
	});

	it('should render without crashing', () => {
		expect(tree.toJSON()).toMatchSnapshot();
	});

	it("should render a '<Spinner />' when loading", () => {
		props.loadingSearch = true;

		component = (
			<Provider>
				<ProductDetail {...props} />
			</Provider>
		);
		tree = renderer.create(component);

		const condition = tree.root.findAllByType(Spinner);
		expect(condition).toHaveLength(1);
	});
});
