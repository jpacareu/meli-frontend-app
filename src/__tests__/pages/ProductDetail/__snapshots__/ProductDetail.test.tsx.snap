// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDetail /> should render without crashing 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
  >
    <ol
      className="breadcrumb-category d-md-flex"
    >
      <li
        aria-current="page"
        className="breadcrumb-category__item"
      >
        Category 1
      </li>
      <li
        aria-current="page"
        className="breadcrumb-category__item"
      >
        Category 2
      </li>
      <li
        aria-current="page"
        className="breadcrumb-category__item__highlight"
      >
        Category 3
      </li>
    </ol>
  </nav>
  <div
    className="item-detail row"
  >
    <div
      className="col-md-7"
    >
      <img
        alt="Title 1"
        className="item-detail__image"
        src="http://mywebsite.com/image.jpg"
      />
    </div>
    <div
      className="col-md-4 col-sm-12"
    >
      <section
        className="item-detail__main-info"
      >
        <div
          className="item-detail__condition-and-sold d-flex"
        >
          <div
            className="item-detail__condition"
          >
            Nuevo
          </div>
          <span>
             - 
          </span>
          <div
            className="item-detail__sold"
          >
            1000
          </div>
        </div>
        <div
          className="item-detail__title"
        >
          Title 1
        </div>
        <div
          className="item-detail__price"
        >
          $ 1.000
        </div>
        <button
          className="item-detail__button"
        >
          Comprar
        </button>
      </section>
    </div>
    <div
      className="col-md-7 col-sm-12"
    >
      <section>
        <summary
          className="item-detail__descripcion-label"
        >
          Descripci√≥n del producto
        </summary>
        <div
          className="item-detail__descripcion-text"
        >
          This is the description
        </div>
      </section>
    </div>
  </div>
</div>
`;

exports[`<ProductDetailDefault /> should render without crashing 1`] = `
<div
  className="spinner__container"
>
  <div
    className="spinner-border"
    role="status"
  >
    <span
      className="sr-only"
    >
      Loading...
    </span>
  </div>
</div>
`;
